crates/c-api/src/val.rs

```diff
@@ -177,7 +177,24 @@ pub struct wasmtime_func_t {
 }
 
 impl wasmtime_val_t {
-    pub fn from_val(cx: impl AsContextMut, val: Val) -> wasmtime_val_t {
+    /// Creates a new `wasmtime_val_t` from a `wasmtime::Val`.
+    ///
+    /// Note that this requires a `RootScope` to be present to serve as proof
+    /// that `val` is not require to be rooted in the store itself which would
+    /// prevent GC. Callers should prefer this API where possible, creating a
+    /// temporary `RootScope` when needed.
+    pub fn from_val(cx: &mut RootScope<impl AsContextMut>, val: Val) -> wasmtime_val_t {
+        Self::from_val_unrooted(cx, val)
+    }
+
+    /// Equivalent of [`wasmtime_val_t::from_val`] except that a `RootScope`
+    /// is not required.
+    ///
+    /// This method should only be used when a `RootScope` is known to be
+    /// elsewhere on the stack. For example this is used when we call back out
+    /// to the embedder. In such a situation we know we previously entered with
+    /// some other call so the root scope is on the stack there.
+    pub fn from_val_unrooted(cx: impl AsContextMut, val: Val) -> wasmtime_val_t {
```

Comment:
Maybe we can rename this `from_val_unscoped` or something? "unrooted" had me convinced that this was passing around `ExternRef::to_raw` results and stuff like that potentially across GCs which is not what's going on.

Reply:
Also, the `wasmtime_val_t::from_val[_unrooted]` constructors never create `Rooted`s so I don't think we actually need two variants of these ones? Its just creating a `Val` (which contains a `Rooted` and means we might be creating a `Rooted` ourselves) that needs the scoping.

Reply:
Good point about the naming, I'll switch to your suggestion.

I thought the same for `from_val` but I ended up realizing that APIs like `wasmtime_global_get` also will want a `RootScope` on them. That's only calling `from_val` but if we forgot a `RootScope` there the any value coming out of a global would forever-after be attached to the store (since the `Rooted` coming out has no other scope). That convinced me to leave the `from_val{,_unscoped}` distinction to continue to encourage the use of `RootScope` anywhere in the C API that might use it.